Fix: propose when creating or signing

katspaugh opened the pull request:
## What it solves
Resolves #3448

## How this PR fixes it
In 3.17.2, before we refactored create/processTransaction, it worked fine.

Process tx:
https://github.com/gnosis/safe-react/blob/release/3.17.2/src/logic/safe/store/actions/processTransaction.ts#L166-L168
https://github.com/gnosis/safe-react/blob/release/3.17.2/src/logic/safe/store/actions/processTransaction.ts#L135-L138

Create tx:
https://github.com/gnosis/safe-react/blob/release/3.17.2/src/logic/safe/store/actions/createTransaction.ts#L167-L169
https://github.com/gnosis/safe-react/blob/release/3.17.2/src/logic/safe/store/actions/createTransaction.ts#L197-L198

Basically, it proposes when signing, or when creating a new tx.

This PR restores that logic.

iamacook commented on src/logic/safe/store/actions/createTransaction.ts:
```diff
@@ -107,20 +107,13 @@ export class TxSender {
   // On transaction completion (either confirming or executing)
   async onComplete(signature?: string, confirmCallback?: ConfirmEventHandler): Promise<void> {
-    const { txArgs, safeTxHash, txProps, dispatch, notifications, isFinalization, approveAndExecute = false } = this
+    const { txArgs, safeTxHash, txProps, dispatch, notifications, isFinalization } = this
```
You can remove `approvedAndExecute` passed to `processTransaction` and all occurances of it throughout the `process`/`createTransaction functions.

iamacook commented on src/logic/safe/store/actions/createTransaction.ts:
```diff
@@ -107,20 +107,13 @@ export class TxSender {
-    // 1) It's a confirmation w/o exection
-    // 2) It's a creation + execution w/o pre-approved signatures
-    if (isOffChainSigning || isImmediateExecution) {
+    if (!isFinalization || !this.txId) {
```
This will throw when a non-owner executes a confirmed transaction.

katspaugh replied:
Wouldn't that be `isFinalization === true`?

katspaugh replied:
I need to test it.

iamacook replied:
I'm not 100% certain. It could be

iamacook reviewed (approved):
As this gets WC approvals 'working', I'll approve this. However, we are over-proposing and errors will appear in the console when non-owners execute.

katspaugh commented:
While testing this, I've encountered some other strange bugs with WC + Safe as a signer.

Need to look more closely into it.

iamacook commented:
> While testing this, I've encountered some other strange bugs with WC + Safe as a signer.

I will revisit the original refactor of these two functions on Friday.

katspaugh commented:
What I found out about WC+Safe as a signer:
* it can create and "sign" transactions, but the signature is empty
* we can still propose these txns to the backend, and it will navigate to the proposed tx
* it will not register a confirmation (because the sig is empty)

The same behavior occurs on 3.17.2, and on this branch (after fixing the proposal conditions).

I've also refactored `processTransaction` a bit more:
* Removed the useless `shouldExecuteTransactionCondition` because it was always returning false. It only makes sense in `createTransaction`, not in `process`.
* Changed the logic around `preApprovingOwner` so that we only send it when only 1 confirmation is missing and send is an owner

iamacook commented on src/logic/safe/store/actions/processTransaction.ts:
```diff
@@ -91,9 +82,10 @@ export const processTransaction = (props: ProcessTransactionArgs): ProcessTransa
-      sigs:
-        generateSignaturesFromTxConfirmations(tx.confirmations, preApprovingOwner) ||
-        getPreValidatedSignatures(sender.from),
+      sigs: generateSignaturesFromTxConfirmations(
```
Why do we no longer need prevalidated signatures? Because it is processing?

katspaugh replied:
Because it's actually called as part of `generateSignaturesFromTxConfirmations`. If confirmations are 0, it will just use the owner address as a sig. So it was just redundant to have this conditional here.

iamacook commented on src/routes/safe/components/Transactions/TxList/modals/ApproveTxModal.tsx:
```diff
@@ -190,21 +190,31 @@ type Props = {
+
   const executionInfo = transaction.executionInfo as MultisigExecutionInfo
-  const thresholdReached = !!(executionInfo && isThresholdReached(executionInfo))
+  const { confirmationsSubmitted, confirmationsRequired } = executionInfo
```
I know it's legacy but we should probably use the type check function here to be safe.

iamacook replied:
Why did you remove the helper function as well? It's only used here?

katspaugh replied:
It was only used here and was just a `>=-comparison`

iamacook reviewed (approved):
There are a lot of cases to check but I trust your logic. I have a few questions, nonetheless

katspaugh commented:
I've tested the following cases:

With MM:
* Creating a tx w/o immediate execution in a 1/1 Safe, executing later as owner
* Creating a tx w/o immediate execution in a 1/1 Safe, executing later as a non-owner
* Creating a tx w/o immediate execution in a 2/3 Safe, second owner confirming and executing
* Creating a tx w/o immediate execution in a 2/3 Safe, second owner confirming, a non-owner executing

With WC and the Safe mobile app as signer (imported EoA keys):
* Creating a tx
* Signing a tx
* Executing – doesn't work, the same was on 3.17.2 (I think this has never been supported?)

With WC and a Safe as signer:
* Creating a tx – proposes but w/o the initial signature (it's empty), shows up in the queue
* Confirming/executing – signature is empty, so essentially doesn't work (same as on 3.17.2)

katspaugh commented:
Yes, basically the fix is that it now proposes a tx on creation, albeit without a signature.

francovenica commented:
I tested regular tx in safes 1/1 and 2/2 and I had no issues creating, executing, rejecting tx with owners and non-owners

I have this safe where the owners are a MM account and another safe
I connect the safe via WC
I was able to create tx with the safe as the owner. The tx is created with 0 signatures at first, but after the tx is successfully executed in the owner safe it changes on its own to have 1 signature already set, so that's working fine
I was able to just confirm tx and the confirm and execute all in the same action

The safe I was testing on
https://pr3455--safereact.review-safe.gnosisdev.com/app/rin:0x978b1681be5e741D888cB8DE076011DE11dE8072/transactions/history

The safe owner rin:0xd7F4a1Cca3428eC09bC0fC9dBDFcDb087803fD59

So far it looks good. The only thing I'm missing to test is with the mobile app.

katspaugh commented:
It's fine, I've tested with the mobile app, it works as good as MM.
